src/pip/_internal/utils/misc.py

```diff
@@ -953,6 +963,36 @@ def _make_build_dir(build_dir):
     write_delete_marker_file(build_dir)
 
 
+class SensitiveText(str):
```

Comment:
My instinct is that this class can be made a lot simpler. Like, couldn't it work just to do this? (no need for `__new__`, `__init__`, `__eq__`, and `__hash__`)

```python
class HiddenText(str):

    def __repr__(self):
        return '<HiddenText {!r}>'.format(str(self))

    def __str__(self):
        return self.redacted

    @property
    def raw(self):
        return str.__str__(self)
```

Then "public" creation functions can be made like this, for example:

```python
def hide_password(password):
    hidden = HiddenText(password)
    hidden.redacted = '****'

    return hidden

a = hide_password('*secret*')

print(str(a))
print(repr(a))
print(a.raw)
```

which gives:

```
****
<HiddenText '****'>
*secret*
```

I think decoupling the creation functions from the class will be better because it keeps the class simpler. Also, after giving it some thought, I think `HiddenText` is a better name in part because it's simpler to type and say.

By the way, your `__hash__()` implementation below I don't think is right because it bases the hash off of the redacted form, which can be the same for different values. The above, on the other hand, does the right thing because it inherits what `str` does, which is to use the underlying string value.

Reply:
> the hash off of the redacted form

That's actually intentional. The idea is that if the redacted form is stored in a hashable location, the redacted form is likely what is being operated upon. The example that I ran into was the urlparse cache where the redacted url was being split. In that case (and I think all cases), the redacted form was what was being cached.

Reply:
Also, the `__eq__` is necessary for proper comparison specifically during tests. I was trying to avoid rewriting all of the assertions which compare actual values to expected values but I'm open to rewriting those tests as well if that's a better approach.

Reply:
> Also, the __eq__ is necessary for proper comparison specifically during tests.

Can you give me an example of some tests that would be affected by what I suggested?

Reply:
This went back to when I was patching the `Link` model but at that point there were a few tests that were failing when comparing the actual url (HiddenText) against the expected url (str).

Here were a couple of examples:
https://github.com/pypa/pip/blob/908e2034fa6b119bfc1a505c9f83140c1d2022a2/tests/unit/test_req_file.py#L564

https://github.com/pypa/pip/blob/908e2034fa6b119bfc1a505c9f83140c1d2022a2/tests/unit/test_req_file.py#L535

Since I'm no longer updating the `Link` model, these tests wouldn't be impacted but if the intent is to move the `HiddenText` cast back into the `Link` model eventually this kind of issue will come up. At that point, it's whether the tests should be updated or the `__eq__` comparison logic.

I defer to your preferences here 😄